File: backend/src/models/trip.py
from __future__ import annotations
from typing import Dict, List, Literal, Optional
from pydantic import BaseModel

TransportMode = Literal["flight", "train", "ferry", "car"]

# Routing modes used by Google Directions / DayItem hops
HopMode = Literal["walking", "driving", "transit"]


class DayItem(BaseModel):
    id: Optional[str] = None
    type: Literal["poi", "hotel", "airport"]
    name: str
    address: str
    lat: float
    lng: float
    duration_mins: Optional[int] = None
    notes: Optional[str] = None
    distance_to_next_km: Optional[float] = None
    travel_time_to_next_mins: Optional[int] = None
    route_polyline_to_next: Optional[str] = None
    # Routing mode for the hop to the next item (walking / driving / transit)
    transport_mode: Optional[HopMode] = None
    photo_url: Optional[str] = None
    photo_urls: Optional[List[str]] = None
    spans_days: int = 1  # >1 for overnight ferry/sleeper train
    # POI detail fields preserved when a POI is added to the itinerary
    rating: Optional[float] = None
    review_count: Optional[int] = None
    price_level: Optional[int] = None
    opening_hours: Optional[str] = None
    indoor_outdoor: Optional[Literal["indoor", "outdoor", "both"]] = None
    nearest_transit: Optional[str] = None
    neighborhood: Optional[str] = None
    booking_required: Optional[bool] = None
    theme_tags: Optional[List[str]] = None
    claude_best_time: Optional[str] = None
    claude_booking_tip: Optional[str] = None
    ai_reason: Optional[str] = None
    busy_times: Optional[Dict[str, List[int]]] = None


class DayPlan(BaseModel):
    day_number: int
    date: str
    leg_number: int
    city: str
    items: List[DayItem] = []


class TripLeg(BaseModel):
    leg_number: int
    origin: str
    destination: str
    departure_date: str
    transport_mode: TransportMode = "flight"
    selected_flight: Optional["FlightOffer"] = None
    flight_booking: Optional["FlightBooking"] = None
    hotel_stays: List["HotelStay"] = []
    days: List[DayPlan] = []
    flight_results: list = []
    hotel_results: list = []


class TripContext(BaseModel):
    home_origin: str
    adults: int
    children: int = 0
    currency: str = "USD"
    legs: List[TripLeg] = []
    unscheduled_pois: List["POI"] = []
    saved_pois: List["POI"] = []


# Resolve forward references after all models are imported
from backend.src.models.flight import FlightOffer, FlightBooking, FlightSearchRequest  # noqa: E402
from backend.src.models.hotel import HotelOffer, HotelStay, HotelBooking, HotelSearchRequest  # noqa: E402
from backend.src.models.poi import POI, POISuggestRequest, DistancesRequest  # noqa: E402

_ns = {
    "FlightOffer": FlightOffer,
    "FlightBooking": FlightBooking,
    "HotelOffer": HotelOffer,
    "HotelStay": HotelStay,
    "HotelBooking": HotelBooking,
    "POI": POI,
    "TripContext": TripContext,
    "DayItem": DayItem,
}

TripLeg.model_rebuild(_types_namespace=_ns)
TripContext.model_rebuild(_types_namespace=_ns)

# Now rebuild cross-module models that reference TripContext or DayItem
FlightSearchRequest.model_rebuild(_types_namespace=_ns)
HotelSearchRequest.model_rebuild(_types_namespace=_ns)
POISuggestRequest.model_rebuild(_types_namespace=_ns)
DistancesRequest.model_rebuild(_types_namespace=_ns)

File: frontend/src/types/trip.ts
// Mirror of all backend Pydantic models

export type DayItemType = "poi" | "hotel" | "airport";

export type TransportMode = "flight" | "train" | "ferry" | "car";

export type POICategory = "activity" | "restaurant" | "sightseeing" | "attraction";

export type HopMode = "walking" | "driving" | "transit";

export type AccommodationType = "hotel" | "ferry_cabin" | "sleeper_train";

export interface DayItem {
  id?: string;
  type: DayItemType;
  name: string;
  address: string;
  lat: number;
  lng: number;
  duration_mins?: number;
  notes?: string;
  distance_to_next_km?: number;
  travel_time_to_next_mins?: number;
  route_polyline_to_next?: string;
  transport_mode?: HopMode;
  spans_days?: number;
}

export interface DayPlan {
  day_number: number;
  date: string;
  leg_number: number;
  city: string;
  items: DayItem[];
}

export interface HotelOffer {
  id: string;
  name: string;
  address: string;
  lat: number;
  lng: number;
  price_per_night: number;
  currency: string;
  rating?: number;
  ai_recommended: boolean;
  ai_reason?: string;
  ai_reason_bullets?: string[];
}

export interface HotelStay {
  hotel: HotelOffer;
  check_in: string;
  check_out: string;
  accommodation_type?: AccommodationType;
}

export interface FlightSegment {
  departure_airport: string;
  arrival_airport: string;
  departure_time: string;
  arrival_time: string;
  duration: string;
  carrier_code: string;
  flight_number: string;
}

export interface FlightOffer {
  id: string;
  price: number;
  currency: string;
  segments: FlightSegment[];
  total_duration: string;
  stops: number;
  ai_recommended: boolean;
  ai_reason?: string;
  ai_reason_bullets?: string[];
}

export interface TripLeg {
  leg_number: number;
  origin: string;
  destination: string;
  departure_date: string;
  transport_mode?: TransportMode;
  origin_airports?: string[];
  destination_airports?: string[];
  selected_flight?: FlightOffer;
  hotel_stays: HotelStay[];
  days: DayPlan[];
  flight_results?: FlightOffer[];
  hotel_results?: HotelOffer[];
}

export interface POI {
  id: string;
  name: string;
  category: POICategory;
  address: string;
  lat: number;
  lng: number;
  opening_hours?: string;
  booking_required: boolean;
  indoor_outdoor?: "indoor" | "outdoor" | "both";
  busy_times?: Record<string, number[]>;
  typical_visit_duration_mins?: number;
  price_level?: number;
  rating?: number;
  review_count?: number;
  photo_url?: string;
  nearest_transit?: string;
  claude_note: string;
  claude_best_time?: string;
  claude_booking_tip?: string;
  ai_recommended?: boolean;
  ai_reason?: string;
  theme_tags: string[];
  neighborhood?: string;
}

export interface SavedHotel {
  id: string;
  item: DayItem;
  day_number: number;
  leg_number: number;
  original_day_index: number;
}

export interface TripContext {
  home_origin: string;
  adults: number;
  children: number;
  currency?: string;
  legs: TripLeg[];
  unscheduled_pois: POI[];
  saved_pois: POI[];
  saved_hotels: SavedHotel[];
}

export interface ItineraryDay {
  day_number: number;
  date: string;
  city: string;
  narrative: string;
  items: DayItem[];
}

export interface TransportSegment {
  leg_number: number;
  mode: TransportMode;
  origin: string;
  destination: string;
  departure_date: string;
  operator?: string;
  duration_mins?: number;
  booking_link?: string;
  booking_ref?: string;
  notes?: string;
}

export interface ExportRequest {
  trip_context: TripContext;
  itinerary: ItineraryDay[];
  transport_segments?: TransportSegment[];
}

export interface RouteSegment {
  distance_km: number | null;
  travel_time_mins: number | null;
  encoded_polyline: string | null;
  mode: HopMode;
}

export interface TripMeta {
  trip_id: string;
  name: string | null;
  is_draft: boolean;
  created_at: string;
  updated_at: string;
}
